### ISSUE TITLE
Representation being assigned to NilClass

### ISSUE THREAD

**Issue by dnd** [apotonick/representable#142]
I'm still trying to figure out what's going on here, but somehow representable is trying to use a representer intended for another class as the representation for `NilClass` in `representable_bindings_for`. It then tries to apply that representation and fails because it's trying to access methods that don't exist on `NilClass`. 

Things start out normally, and then an action in my app will fail, and from that point on `NilClass` is mapped to that representer. I utilize representers as `MyRepresenter.represent(object)`. I don't think I'm doing anything crazy, but I haven't been able to track down the cause of the problem yet.

Even though I don't have much information on this yet, I'm putting it out there in case anyone else has run across it.

**Comment by dnd** [comment]
Here's a gist which demonstrates the failure:

https://gist.github.com/dnd/1376e015c0b4a4435558

I don't know exactly what role Virtus plays in the whole thing, because it all works fine without the initial attempt to represent a nil object, but it does have to do with Virtus having a `nil` default. It's easy enough to stop this with a nil check, but the persisting effect to future calls seems like it's probably a bug in representable. I'm not sure where it's all happening yet though.

**Comment by apotonick** [comment]
Thanks for your test, made it really simple to follow.

The explanation is very simple, too.

1. You extend `nil` with your Representer (works).
2. After `nil` got extended, the `Representable::extended` hook is run, trying to build `representable_attrs` in `nil`. This fails.
3. However, `nil` is now marked as extended by Ruby, because you catch the exception and do not handle it properly.
4. Next time you try to represent `nil`, the `extended` is not run, because Ruby thinks `nil` is "representable" already. It assumes all is good to go and crashes, because `nil` is in some stale state.

Solution: Please don't catch exceptions when representing but check what you pass into the representer.

**Comment by dnd** [comment]
The exception was caught by a default rescue for the API so that it can return a 500 error in event of failure. What could be done to handle the exception instead, because once the exception is thrown it's too late.

I do realize that attempting to represent nil was the root cause, but should that even be allowed to happen? Is there a legitimate reason for being able to extend and represent nil? For the most part none of the places I attempt to represent are able to return nil, but obviously this one slipped through. When this one instance slips throug it breaks the application for the entirety of its life. It's also not very obvious when the error is occurring as to why it's actually occurring, and could be quite difficult to track down where it's originating.

If there isn't a good use case for representing nil, then should representable maybe prevent the extension on `nil` from even happening?

**Comment by apotonick** [comment]
@dnd Can you post an example of how you treat this edge-case in your application code? Surely, that'll help others! :heart:

**Comment by andrewdang17** [comment]
how did you fix this @dnd? I'm having this issue too

**Comment by dnd** [comment]
I thought I had responded to this, but to give everyone the workaround in case anyone else runs into it, I ended up checking for nil before any place that I attempted to represent something. For me, that was wrapped up in only one or two locations for the entire app, but for others it might be different. 

If nothing has changed in Representable I still think that as breaking as it is that it's a failing of the Representable code to even attempt representation of `nil`. I can't think of a valid use case for doing so, and including a simple check in the Representable code would save every user of the library that potential headache.

If it has changed, then yay! :)
